src/nxt_socket.c

```diff
@@ -285,6 +285,20 @@ nxt_socket_close(nxt_task_t *task, nxt_socket_t s)
     nxt_err_t   err;
     nxt_uint_t  level;
 
+#if (NXT_HAVE_UNIX_DOMAIN)
+    socklen_t           len;
+    struct sockaddr_un  sa = {0};
+
+    len = sizeof(struct sockaddr_un) - 1;
+    if ((getsockname(s, (struct sockaddr *)&sa, &len) == 0)
+        && (sa.sun_family == AF_UNIX))
+    {
+        ((char *) &sa)[len] = '\0';
+        nxt_debug(task, "socket unlink(%s)", sa.sun_path);
+        (void) unlink(sa.sun_path);
```

Comment:
Right, the problem here is that it seems to be the router process doing the unlink(2) and when unit is running as root the router process is running as an unprivileged user (nobody here).

I wonder if it would be possible to setup a callback function like there is for the isolation or cgroup stuff for cleaning things up when an application is terminated. The socket is created in the main unit process so hopefully it still has the details...

Or chown(2) the socket to the unprivileged user... the sockets are being created 666 regardless (here anyway).

Reply:
Re: callback:  If the file is owned by root, then only the main process can remove it.  I thought about sending the sfd back to the main process so that it's the main process the one that unlink(2)s and close(2)s it.  But that's not an easy thing.  I'd prefer to avoid that.

Re: chown(2):  Hmm, I like the idea.  If it's currently 666 I guess there's no added insecurity by chown(2)ing it to nobody.  The directory permissions are a problem, though.

Reply:
Re: callback:  If the file is owned by root, then only the main process can remove it.  I thought about sending the sfd back to the main process so that it's the main process the one that unlink(2)s and close(2)s it.  But that's not an easy thing.  I'd prefer to avoid that, if we can.

Re: chown(2):  Hmm, I like the idea.  If it's currently 666 I guess there's no added insecurity by chown(2)ing it to nobody.  The directory permissions are a problem, though.

I don't know...  is there any way to tell the kernel: "I want to unlink this file, but don't remove it from the filesystem until the last fd has been closed"?

Maybe the answer to that is using abstract sockets in Linux, and leaving other systems with non-unlinked sockets?

Reply:
I guess you referred to abstract sockets the other day, right?  For anyone else reading this, see unix(7):

```
DESCRIPTION
       [...]
   Address format
       [...]

       Three  types  of  address  are distinguished in the sock‐
       addr_un structure:

       [...]

       *  abstract: an abstract socket address is  distinguished
          (from  a pathname socket) by the fact that sun_path[0]
          is a null byte ('\0').  The socket’s address  in  this
          namespace is given by the additional bytes in sun_path
          that  are  covered  by the specified length of the ad‐
          dress structure.  (Null bytes in the name have no spe‐
          cial significance.)  The name has no  connection  with
          filesystem pathnames.  When the address of an abstract
          socket  is  returned,  the returned addrlen is greater
          than sizeof(sa_family_t) (i.e., greater than  2),  and
          the  name of the socket is contained in the first (ad‐
          drlen - sizeof(sa_family_t)) bytes of sun_path.

       [...]

   Abstract sockets
       Socket permissions have no meaning for abstract  sockets:
       the  process  umask(2)  has no effect when binding an ab‐
       stract socket, and changing the ownership and permissions
       of the object (via fchown(2) and fchmod(2)) has no effect
       on the accessibility of the socket.

       Abstract sockets automatically disappear  when  all  open
       references to the socket are closed.

       The  abstract socket namespace is a nonportable Linux ex‐
       tension.

[...]

VERSIONS
       SCM_CREDENTIALS  and  the  abstract namespace were intro‐
       duced with Linux 2.2 and should not be used  in  portable
       programs.  (Some BSD‐derived systems also support creden‐
       tial passing, but the implementation details differ.)
```

Reply:
> Re: callback: If the file is owned by root, then only the main process can remove it. I thought about sending the sfd back to the main process so that it's the main process the one that unlink(2)s and close(2)s it. But that's not an easy thing. I'd prefer to avoid that, if we can.

Yes, what I was hoping was that you might be able to do something like what the isolation and cgroup code do by hooking into the nxt_main_process_cleanup() function for when an application terminates. This is run from the main unit process. E.g

```
$ sudo /opt/unit/sbin/unitd --control 127.0.0.1:8080 --no-daemon
2022/07/20 17:13:27 [info] 5220#5220 unit 1.28.0 started
^C[5220] Cleaning up cgroups for (5255/"sqroot" prototype)
[5220] Cleaning up cgroups for (5224/"add" prototype)
[5220] Cleaning up cgroups for (5247/"divide" prototype)
[5220] Cleaning up cgroups for (5229/"multiply" prototype)
[5220] Cleaning up cgroups for (5226/"subtract" prototype)
```

In the socket case I guess you'd want to do it at reconfigure/shutdown time. Now I don't know if a facility already exists or if it might not be too much work to add one, just another idea...

> I don't know... is there any way to tell the kernel: "I want to unlink this file, but don't remove it from the filesystem until the last fd has been closed"?

Not as far as I know, the closest paradigm I can think of is the open file, unlink it (so it **can't** be easily accessed) and then close. But of course that doesn't help here...

> Maybe the answer to that is using abstract sockets in Linux, and leaving other systems with non-unlinked sockets?

Heh, that would certainly nicely solve the problem on Linux!